reactor-core/src/main/java/reactor/core/publisher/SinkManyEmitterProcessor.java

```diff
@@ -383,10 +383,12 @@ public Object scanUnsafe(Attr key) {
 	}
 
 	final void drain() {
-		if (WIP.getAndIncrement(this) != 0) {
+		if (WIP.get(this) != 0) {
```

Comment:
In the cases where more than 1 thread calling the drain function, all the threads are incrementing the WIP counter while only one of them proceeds to clear the queue and decrements the counter. This leaves the WIP in unclean state.

Reply:
By separating the atomic operation into two operations you break the exclusive access to the critical section below. Please start with creating JCStress tests if you're considering to spend more time on this. These drain methods are quite a critical piece and ensuring proper lock-free coordination between different actors is essential. Another aspect is performance so that we make as little volatile accesses as we can.